docs/proposals/flannel-integration.md

```diff
@@ -0,0 +1,165 @@
+    link/ether 42:01:0a:f0:00:04 brd ff:ff:ff:ff:ff:ff
+
+node2 $ ip -d link show flannel.1
+4: flannel.1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1410 qdisc noqueue state UNKNOWN mode DEFAULT
+    link/ether 56:71:35:66:4a:d8 brd ff:ff:ff:ff:ff:ff
+    vxlan
+node2 $ ip -d link show eth0
+2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc mq state UP mode DEFAULT qlen 1000
+    link/ether 42:01:0a:f0:00:03 brd ff:ff:ff:ff:ff:ff
+```
+
+Note that we're ignoring cbr0 for the sake of simplicity. Spin-up a container on each node. We're using raw docker for this example only because we want control over where the container lands:
+
+```
+node1 $ docker run -it radial/busyboxplus:curl /bin/sh
+[ root@5ca3c154cde3:/ ]$ ip addr show
+1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue
+8: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1410 qdisc noqueue
+    link/ether 02:42:12:10:20:03 brd ff:ff:ff:ff:ff:ff
+    inet 192.168.32.3/24 scope global eth0
+       valid_lft forever preferred_lft forever
+
+node2 $ docker run -it radial/busyboxplus:curl /bin/sh
+[ root@d8a879a29f5d:/ ]$ ip addr show
+1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue
+16: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1410 qdisc noqueue
+    link/ether 02:42:12:10:0e:07 brd ff:ff:ff:ff:ff:ff
+    inet 192.168.14.7/24 scope global eth0
+       valid_lft forever preferred_lft forever
+[ root@d8a879a29f5d:/ ]$ ping 192.168.32.3
+PING 192.168.32.3 (192.168.32.3): 56 data bytes
+64 bytes from 192.168.32.3: seq=0 ttl=62 time=1.190 ms
+```
+
+__What happened?__:
+
+From 1000 feet:
+* vxlan device driver starts up on node1 and creates a udp tunnel endpoint on 8472
+* container 192.168.32.3 pings 192.168.14.7
+    - what's the MAC of 192.168.14.0?
+        - L2 miss, flannel looks up MAC of subnet
+        - Stores `192.168.14.0 <-> 56:71:35:66:4a:d8` in neighbor table
+    - what's tunnel endpoint of this MAC?
+        - L3 miss, flannel looks up destination VM ip
+        - Stores `10.240.0.3 <-> 56:71:35:66:4a:d8` in bridge database
+* Sends `[56:71:35:66:4a:d8, 10.240.0.3][vxlan: port, vni][02:42:12:10:20:03, 192.168.14.7][icmp]`
+
+__But will it blend?__
+
+Kubernetes integration is fairly straight-forward once we understand the pieces involved, and can be prioritized as follows:
+* Kubelet understands flannel daemon in client mode, flannel server manages independent etcd store on master, node controller backs off cidr allocation
+* Flannel server consults the Kubernetes master for everything network related
+* Flannel daemon works through network plugins in a generic way without bothering the kubelet: needs CNI x Kubernetes standardization
+
+The first is accomplished in this PR, while a timeline for 2. and 3. are TDB. To implement the flannel api we can either run a proxy per node and get rid of the flannel server, or service all requests in the flannel server with something like a go-routine per node:
+* `/network/config`: read network configuration and return
```

Comment:
Recent refactoring in flannel made storage less dependent on etcd. It should definitely be possible to add another implementation that is backed by k8s api-server instead of etcd. On the one hand, I'm not a fan of flannel depending on k8s as these projects should be orthogonal to each other. But from pragmatic point of view, it is something that I think I can live with.